### PR TITLE
Enable Go Modules

### PR DESCRIPTION
**Pull request by gdbelvin**
Turn on go modules for certificate-transaprency-go

* All dependencies are now pinned at a particular revision.
* Including all references to the Trillian repository
    - `go generate -run protoc` references pinned proto files in Trillian . 
    - `{TRILLIAN_PATH}/scripts/` refer to the pinned Trillian version.
 * Monolog 
    - `"${MONO_PATH}/scripts` refer to the pinned to the Monolog repo. 
* Vendored files will be been removed in a subsequent PR
* golangci-lint runs slower in module mode so we only run it when the `-race` is not enabled.

### CODE REVIEW

**Comment by gdbelvin** [comment]
Depends on google/trillian#1668

**Comment by gdbelvin** [comment]
Depends on https://github.com/google/trillian/pull/1673

File: .travis.yml [hunk 1]
```diff
@@ -7,14 +7,21 @@ go_import_path: github.com/google/certificate-transparency-go
 
 
 env:
+  global:
+  - GO111MODULE=on
+  matrix:
   - GOFLAGS=
-  - GOFLAGS=-race
+  - GOFLAGS=-race NO_LINTER=true
```
Comment (AlCutter) [inline .travis.yml]: Perhaps we only need to only run the linter in the base case? (is there any code which would be linted by `WITH_COVERAGE=true` but not in the empty `GOFLAGS` case?)

File: .travis.yml [hunk 2]
```diff
@@ -28,26 +35,23 @@ install:
...
 before_script:
-  - ${GOPATH}/src/github.com/google/trillian/scripts/resetdb.sh --force
+  - bash $(go list -f '{{.Dir}}' github.com/google/trillian)/scripts/resetdb.sh --force
   - ./scripts/resetctdb.sh --force
 
 script:
   - set -e
   - cd $HOME/gopath/src/github.com/google/certificate-transparency-go
-  - ./scripts/presubmit.sh ${PRESUBMIT_OPTS} ${WITH_COVERAGE:+--coverage}
+  - ./scripts/presubmit.sh ${PRESUBMIT_OPTS} ${WITH_COVERAGE:+--coverage} ${NO_LINTER:+--no-linters}
```
Comment (AlCutter) [inline .travis.yml]: This expansion may cause surprise in the future - it'll disable linters even if someone sets `NO_LINTER=false` above... how about doing something like:

```bash
   - if [ "${NO_LINTER}" == "true" ]; then export PRESUBMIT_OPTS="${PRESUBMIT_OPTS} --no-linters"; fi
   - ./scripts/presub...
```
?

File: trillian/integration/ct_functions.sh [hunk 3]
```diff
@@ -37,7 +37,7 @@ ct_prep_test() {
     "${CT_GO_PATH}/scripts/resetctdb.sh" --force
     # Wipe the incident database
     echo "Wiping and re-creating incident database"
-    "${MONO_PATH}/scripts/resetmondb.sh" --force
+    bash "${MONO_PATH}/scripts/resetmondb.sh" --force
```
Comment (AlCutter) [inline trillian/integration/ct_functions.sh]: Probably should fix this upstream too (and then later remove the workaround)?

File: .travis.yml (hunk 1 above)
Comment (gdbelvin) [reply .travis.yml]: I can't think of any value added by running the linter with coverage on

File: trillian/integration/ct_functions.sh (hunk 3 above)
Comment (gdbelvin) [reply trillian/integration/ct_functions.sh]: By upstream do you mean https://github.com/golang/go/issues/27455 ?
I've already fixed it in the Trillian repo here: https://github.com/google/trillian/pull/1673

File: .travis.yml (hunk 2 above)
Comment (gdbelvin) [reply .travis.yml]: I'll use PRESUBMIT_OPTS

File: trillian/integration/ct_functions.sh (hunk 3 above)
Comment (AlCutter) [reply trillian/integration/ct_functions.sh]: Oh, it's "WAI" for go mod? Readonly seems sensible, but !x seems sad. Oh well.

Comment (RJPercival) [reply trillian/integration/ct_functions.sh]: I don't think it's WAI, based on https://github.com/golang/go/issues/31481#issuecomment-500415155. However, until they fix it, we have to workaround it.

**Pull request merged**